frontend/public/components/environment.jsx

```diff
@@ -44,212 +45,215 @@ export const envVarsToArray = (initialPairObjects) => {
+      currentEnv[i] = env.nameValuePairs;
+      const modified = !_.isEqual(currentEnv, envVarsToArray(rawEnvData));
+      this.setState({
+        currentEnvVars: currentEnv,
+        success: null,
+        modified,
+      });
+    }
 
-
-  /**
-   * Make it so. Patch the values for the env var changes made on the page.
-   * 1. Validate for dup keys
-   * 2. Throw out empty rows
-   * 3. Use add command if we are adding new env vars, and replace if we are modifying
-   * 4. Send the patch command down to REST, and update with response
-   *
-   * @param e
-   */
-  _saveChanges(e) {
-    const {envPath, rawEnvData, obj} = this.props;
-    const {currentEnvVars} = this.state;
-    e.preventDefault();
-
-    // Convert any blank values to null
-    const kind = modelFor(obj.kind);
-
-    const patch = currentEnvVars.map((finalPairsForContainer, i) => {
-      let op = 'add';
-      const path = _.isArray(rawEnvData) ? `/${envPath.join('/')}/${i}/env` : `/${envPath.join('/')}/env`;
-      if (_.isArray(rawEnvData)) {
-        if (rawEnvData[i].env) {
-          op = 'replace';
-        }
-      } else {
-        if (rawEnvData.env) {
-          op = 'replace';
-        }
-      }
-      return {path, op, value: this._envVarsToNameVal(finalPairsForContainer)};
-    });
-    const promise = k8sPatch(kind, obj, patch);
-    this.handlePromise(promise).then((res) => {
-      const newEnvData = _.get(res, envPath);
+    /**
+     * Reset the page to initial state
+     * @private
+     */
+    _reload() {
+      const {rawEnvData} = this.props;
       this.setState({
-        success: 'Successfully updated the environment variables.',
+        currentEnvVars: envVarsToArray(rawEnvData),
         errorMessage: null,
-        currentEnvVars: envVarsToArray(newEnvData),
+        success: null,
         modified: false,
         stale: false
       });
-    });
-  }
+    }
+
+    /**
+     * Build out our currentEnvVars state object from our incoming props.
+     * If there is a change and are read/write let the user know we have updated vars otherwise just refresh the page.
+     * For no change return null
+     *
+     * @param nextProps
+     * @param prevState
+     */
+    static getDerivedStateFromProps(nextProps, prevState) {
+      const { currentEnvVars } = prevState;
+      const { rawEnvData, readOnly } = nextProps;
+      const incomingEnvVars = envVarsToArray(rawEnvData);
+      if (_.isEqual(incomingEnvVars, currentEnvVars)) {
+        return null;
+      }
+      return readOnly ? { currentEnvVars: incomingEnvVars } : { stale: true, success: null };
+    }
+
 
-  render() {
-    const {errorMessage, success, inProgress, currentEnvVars, stale, configMaps, secrets} = this.state;
-    const {rawEnvData, readOnly, obj} = this.props;
 
-    if (!configMaps || !currentEnvVars || !secrets) {
-      return <LoadingBox />;
+    /**
+     * Make it so. Patch the values for the env var changes made on the page.
```

Comment:
`Make it so.`

![screen shot 2018-07-24 at 10 17 29 am](https://user-images.githubusercontent.com/280512/43144317-d38770e0-8f2a-11e8-98e8-a9601e350197.png)

Couldn't resist 😄